Mknr/plot routine

EtherealMC-Bit opened the pull request:
Post-processing plot routine to plot field diagnostics during or after runs.

EtherealMC-Bit committed:
Copied ArrayPlot and utility functions over

Currently plotting does not work

EtherealMC-Bit committed:
Moved plotting utilities into plotting.py

EtherealMC-Bit committed:
Fixed issue with ghost cells, saved plot file

EtherealMC-Bit committed:
Changed to Peter's plot parameters

KZhu-ME commented on mewarpx/examples/capacitive_discharge_copy.py:
```diff
@@ -0,0 +1,237 @@
+"""
```
If this file is going to be committed/pushed to the remote repo, you should give it a more descriptive name.

KZhu-ME reviewed (commented):
Everything looks good so far!

peterscherpelz commented:
@KZhu-ME Yeah I don't think we need to stay with a file in `examples`. The card has:

> Success for this card requires a test that shows a WarpX run with plots automatically made at every diagnostic step and a case where all the plots are generated after the run completes.

I think that getting those tests working will be sufficient as demonstrations too.

EtherealMC-Bit committed:
Added functionality to plot at diag steps and after simulation end

EtherealMC-Bit committed:
Removed old debug print outs

roelof-groenewald commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -1181,7 +1181,7 @@ def get_gathered_charge_density_fp(level):
     '''
 
-    return _get_mesh_field_list(libwarpx.warpx_getGatheredChargeDensityFP, level, None, True)
+    return _get_mesh_field_list(libwarpx.warpx_getGatheredChargeDensityFP, level, None, False)
```
This will break the direct solver. The solution used for `get_gathered_phi_fp` is definitely a better way to go.

roelof-groenewald commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -1210,7 +1210,7 @@ def get_mesh_phi_fp(level, include_ghosts=True):
 
 
-def get_gathered_phi_fp(level):
+def get_gathered_phi_fp(level, include_ghosts=True):
```
We should remove this entire wrapper as well as the `get_gathered_rho_fp` wrapper. We can call `_get_mesh_field_list(libwarpx.warpx_getGatheredChargeDensityFP, level, None, include_ghosts)` directly from mwxrun.py.

roelof-groenewald commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,661 @@
+    """
+
+    def __init__(self, nxyz, pos_lims, geom, dt, periodic=True):
+        self.nxyz = nxyz
+        self.pos_lims = pos_lims
+        self.geom = geom
+        self.dt = dt
```
We should be able to get all this information from importing the `mwxrun` object.

roelof-groenewald reviewed (changes requested):
I left some comments from a quick preliminary review.

KZhu-ME commented on mewarpx/mewarpx/util.py:
```diff
@@ -4,15 +4,13 @@
 mewarpx_dir = os.path.dirname(os.path.abspath(
     inspect.getfile(inspect.currentframe())))
 
-
```
I believe it's standard to have 2 blank lines between non-class functions, this includes between the header imports and the first function definition.

EtherealMC-Bit replied:
yep definitely an accident, nice catch

KZhu-ME reviewed (approved):
Overall, looks good, only 1 minor comment

peterscherpelz commented:
@roelof-groenewald @MKieburtz Regarding the issue of getting nx/ny/nz etc. definitely getting them from mwxrun on an initialized run is good; what we haven't answered is how to store those parameters if we're instead post-processing a run from a yt file where we haven't initialized an mwxrun object.

Maybe we use the yt data to initialize the mwxrun object? I'm really not sure at this point...

EtherealMC-Bit committed:
Removed some wrappers in favor of directly calling from mwxrun

EtherealMC-Bit committed:
Removed unneeded SimInfo class, instead use mwxrun

roelof-groenewald commented:
> @roelof-groenewald @MKieburtz Regarding the issue of getting nx/ny/nz etc. definitely getting them from mwxrun on an initialized run is good; what we haven't answered is how to store those parameters if we're instead post-processing a run from a yt file where we haven't initialized an mwxrun object.
> 
> Maybe we use the yt data to initialize the mwxrun object? I'm really not sure at this point...

Perhaps the easiest will be to have the possibility to pass a yt plotfile to that function - if one is given nx/ny/nz etc. will be pulled directly from the plotfile; if not mwxrun will be used. How does that sound?

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -183,16 +183,16 @@ def get_npart_species_dict(self):
 
-    def get_rho_grid(self):
+    def get_rho_grid(self, include_ghosts=True):
         """Get rho segments on the grid for each tile of each processor.
 
         Returns:
             A list of numpy arrays, the list has an array for every tile and
             each array has dimensions given by the number of cells in that tile.
         """
-        return _libwarpx.get_mesh_charge_density_fp(self.lev)
+        return _libwarpx._get_mesh_field_list(_libwarpx.libwarpx.warpx_getChargeDensityFP, self.lev, None, include_ghosts)
```
Do we actually use this at all? If not we should just remove it. If we find we ever do want it we can put it back then.

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -201,26 +201,26 @@ def get_gathered_rho_grid(self):
 
         """
-        return _libwarpx.get_gathered_charge_density_fp(self.lev)
+        return _libwarpx._get_mesh_field_list(_libwarpx.libwarpx.warpx_getGatheredChargeDensityFP, self.lev, None, include_ghosts)
```
```suggestion
        return _libwarpx._get_mesh_field_list(
            _libwarpx.libwarpx.warpx_getGatheredChargeDensityFP, 
            self.lev, None, include_ghosts
        )
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -201,26 +201,26 @@ def get_gathered_rho_grid(self):
 
-    def get_phi_grid(self):
+    def get_phi_grid(self, include_ghosts=True):
         """Get phi segments on the grid for each tile of each processor.
 
         Returns:
             A list of numpy arrays, the list has an array for every tile and
             each array has dimensions given by the number of cells in that tile.
         """
-        return _libwarpx.get_mesh_phi_fp(self.lev)
+        return _libwarpx._get_mesh_field_list(_libwarpx.libwarpx.warpx_getPhiFP, self.lev, None, include_ghosts)
```
Same thing here; I don't think we use the non-gathered grid functions anywhere so if that is the case we should just remove it from mwxrun.py.

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -201,26 +201,26 @@ def get_gathered_rho_grid(self):
             multifab object is returned on processors other than root.
         """
-        return _libwarpx.get_gathered_phi_fp(self.lev)
+        return _libwarpx._get_mesh_field_list(_libwarpx.libwarpx.warpx_getGatheredPhiFP, self.lev, None, include_ghosts)
```
```suggestion
        return _libwarpx._get_mesh_field_list(
            _libwarpx.libwarpx.warpx_getGatheredPhiFP, 
            self.lev, None, include_ghosts
        )
```

roelof-groenewald commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,640 @@
+
+    for param in what_to_plot:
+        if param == 'rho':
+            data = np.array(mwxrun.get_gathered_rho_grid()[0])
```
Should this now be:
```suggestion
            data = np.array(mwxrun.get_gathered_rho_grid(include_ghosts=False)[0])
```

roelof-groenewald reviewed (changes requested):
Thanks for moving those functions from _libwarpx.py. I left a few more comments. This is looking good.

EtherealMC-Bit committed:
Removed unused get non gathered rho and phi functions

EtherealMC-Bit committed:
fixed long lines in mwxrun

EtherealMC-Bit committed:
Include ghosts in get_gathered_phi() for plotting

EtherealMC-Bit committed:
Revert "Include ghosts in get_gathered_phi() for plotting"

This reverts commit 280df772c96044c61ddc88270e603eed9e22bb41.

EtherealMC-Bit committed:
Don't include ghosts when plotting rho

EtherealMC-Bit committed:
Removed 2 single line wrappers and readded upstream function

EtherealMC-Bit committed:
Merge branch 'memaster' into mknr/plot_routine

EtherealMC-Bit committed:
Readded line breaks in the get_mesh_charge_density_fp comment

nruof commented on mewarpx/examples/diagnostic_plotting.py:
```diff
@@ -0,0 +1,196 @@
+mwxrun.simulation.solver = solver
+mwxrun.simulation.time_step_size = DT
+mwxrun.simulation.max_steps = max_steps
+
+mwxrun.simulation.add_species(
+    electrons,
+    layout = picmi.GriddedLayout(
+        n_macroparticle_per_cell=number_per_cell_each_dim, grid=grid
+    )
+)
+mwxrun.simulation.add_species(
+    ions,
+    layout = picmi.GriddedLayout(
+        n_macroparticle_per_cell=number_per_cell_each_dim, grid=grid
+    )
+)
```
Use the new mepicmi Species, which adds species to the simulation by default, without having to specify add_species separately.

peterscherpelz replied:
I think you pass `n_macroparticle_per_cell` and `grid` in the initial `mepicmi.Species` calls and then you can delete these `add_species` blocks entirely?

EtherealMC-Bit committed:
Used mepicmi.Species instead of picmi.Species

EtherealMC-Bit committed:
Revert 'Used mepicmi.Species instead of picmi.Species'

EtherealMC-Bit committed:
Correctly used mepicmi species (import after init_libwarpx)

EtherealMC-Bit committed:
Added field diagnostic wrapper

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+        "draw_cbar": True,
+        "draw_surface": False,
+        "draw_image": False,
+        "draw_contourlines": True,
```
I think we both agree these are better defaults, so let's go ahead and change them now.
```suggestion
        "draw_image": True,
        "draw_contourlines": False,
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+            'image': {
+                'draw_image': True,
+                'default_ticks': True,
+                'draw_contourlines': False,
```
These would now be duplicates.
```suggestion
                'default_ticks': True,
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+            'titlesize': 'large',
+            'cmap': 'viridis',
+            'draw_image': True,
+            'draw_contourlines': False,
```
These would now be duplicates.
```suggestion
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+            'titlesize': 'large',
+            'cmap': 'viridis',
+            'draw_image': True,
+            'draw_contourlines': False,
```
These would now be duplicates.
```suggestion
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+            points (float or list of floats): If plot1d is True, position(s) (in
+                m) to plot. If None, plot equally spaced positions. If this is
+                used, it overrides numpoints.
+            draw_contourlines (bool): If True, draw contour lines. Default True.
```
```suggestion
            draw_contourlines (bool): If True, draw contour lines. Default False.
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+                Default False.
+            draw_image (bool): If True, use ax.imshow() rather than
+                ax.contourf. Displays pixelation with no interpolation.
+                Default False.
```
```suggestion
                Default True.
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+        self.dim = len(self.array.shape)
+        xaxis_idx, yaxis_idx, sliceaxis_idx, self.sliceaxis_str = (
+            get_axis_idxs(self.params["xaxis"], self.params["yaxis"],
+                               self.dim))
```
```suggestion
                          self.dim)
        )
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+axis_dict_2d = {'r': 0, 'x': 0, 'z': 1}
+
+
+def return_iterable(x, depth=1):
```
Let's move this function into util. It is a pretty general function that we're likely to use at some other point in there.

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+                self.params["numpoints"])
+        idx_list = []
+        barrier_indices = []
+        for point in return_iterable(self.params["points"]):
```
```suggestion
        for point in util.return_iterable(self.params["points"]):
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+                               linewidth=1, color="blue", arrowstyle='->',
+                               arrowsize=1.5)
+
+        #plt.show()
```
```suggestion
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+    return dslice
+
+
+def plot_parameters(what_to_plot, plot_name, diag_step=None):
```
Let's move this into the field diagnostic class/file. It doesn't really make sense outside of it that I can tell. Please also add docstring.

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -0,0 +1,55 @@
+class FieldDiagnostic(WarpXDiagnostic):
+    def __init__(self, diag_steps, diag_data_list, grid, name, write_dir,
+                 plot_on_diag_step=False, plot_data_list=None,**kwargs):
+        """
```
docstring should be expanded: (1) Summary description of diagnostic before going into arguments. (2) indicate what valid strs for data_list are (or where a list ov valid strs) can be found (3) Describe other input arguments.

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -0,0 +1,633 @@
+    # There is one more point on the grid than cell number
+    return np.linspace(xmin, xmax, npts + 1)
+
+def recursive_update(d, u):
```
Let's move this one to util too. It's incidental to plotting and not unlikely we'll use it elsewhere in the future too.

peterscherpelz commented on mewarpx/examples/diagnostic_plotting.py:
```diff
@@ -0,0 +1,190 @@
+##########################
+
+field_diag = field_diagnostic.FieldDiagnostic(
+    name = 'diags',
```
No spaces around `=` within a function call

peterscherpelz commented on mewarpx/examples/diagnostic_plotting.py:
```diff
@@ -0,0 +1,190 @@
+# --- Number of time steps
+max_steps = int(TOTAL_TIME / DT)
+diag_steps = 2
+diagnostic_intervals = "::1"
```
Is this used anywhere now?

peterscherpelz commented on mewarpx/examples/diagnostic_plotting.py:
```diff
@@ -0,0 +1,190 @@
+# simulation run
+##########################
+
+mwxrun.simulation.step()
```
Only go for a set number of steps?

peterscherpelz reviewed (commented):
Just lots of smaller changes / suggestions. Everything looks to be on a good track. Main things I see are implementing actual unit tests, and supporting plotting directly from yt files. Thanks!

KZhu-ME reviewed (commented):
Aside from all the current comments, everything looks good! Will there be a unit test for this feature?

EtherealMC-Bit committed:
Requested changes and refactoring

nruof committed:
Nruof/particle diag wrap (#30)

* Python - add new particle attributes at runtime (#2112)

* exposes AddRealComp to Python to allow extra particle attributes to be added at runtime; also includes a new function to grab a particle data array from the name of the component rather than the index

* added function to get index of a particle component given the PID name

* changed new get component index and get_particle_arrays_from_comp_name functions to take species name as argument rather than species id

* changed warpx_addRealComp to accept a species name as input and only add the new component for that species

* Use new boundary input (#2110)

* support only new boundary interface. set pml flags to 0 as default

* use new boundary interface in Examples/Modules/

* use new boundary interface in the input for tests in Examples/Physical_applications

* use new boundary interface in the input for tests in Examples/Tests

* use new bc

* fix typo

* fix typo

* Update Examples/Tests/SingleParticle/inputs_2d

* fix bc in input

* specific bc for electrostatic inputs

* resetting benchmark for PEC particle CI test after fixing the default pml flags from 1 to 0

* Update Examples/Physics_applications/laser_acceleration/inputs_2d_rz

* remove documentation on previous boundary input

* Comment to clarify where in the code is_periodic is appended and how it is defined

* add default parameters for BC in doc

* Implements a periodically repeating plasma lenses (#2080)

* Implements a periodically repeating plasma lens

* Added documentation for plasma lenses

* Added m_n_lenses to avoid use of size in device code

* Change arrays to device arrays

* Put DeviceVectors in the MultiParticleContainer

* Update Source/Particles/MultiParticleContainer.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* Update Source/Particles/Gather/GetExternalFields.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* Added CI test

* Updated CI test

* LGTM clean up

* Moved literal namespace inside the routine

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* Use only species name in libwarpx (#2119)

* changed all particle interacting functions in libwarpx to use the species name rather than id, also changed the functions to get particle array data to use the component name rather than index

* removed unneeded function declaration

* AMReX/PICSAR: Weekly Update (#2107)

Update AMReX to AMReX-Codes/amrex@ee8facf7e09d0b40af1fe680665660c24a19d32a

No changes im PICSAR since last update.

```
./Tools/Release/updatePICSAR.py
./Tools/Release/updateAMReX.py
```

* ChecksumAPI: yt 4+ ds.force_periodicity() (#2117)

* ChecksumAPI: yt 4+ ds.force_periodicity()

Mitigate issues of the form
```
RuntimeError: yt attempted to read outside the boundaries of a non-periodic domain along dimension 0.
Region left edge = -7.5e-06 code_length, Region right edge = 7.500000000000002e-06 code_length
Dataset left edge = -7.5e-06 code_length, Dataset right edge = 7.5e-06 code_length

This commonly happens when trying to compute ghost cells up to the domain boundary. Two possible solutions are to select a smaller region that does not border domain edge (see https://yt-project.org/docs/analyzing/objects.html?highlight=region)
or override the periodicity with
ds.force_periodicity()
```

* make yt3 compatible

* simplify code (#2108)

* assert that the NCI corrector is used with Esirkepov only (#2114)

* Update picmi.py for new injection_style default (#2109)

* AMReX: ROCMClang CMake (#2120)

* AMReX: 91fa2b7e6ff33117bd537635c8c55192359863a2

Include HIP ROCMClang recog. in AMReX:
  https://github.com/AMReX-Codes/amrex/pull/2184

* HIP CI: Use clang++ directly

Looks like the C++14 flags are missing

* Multi-J Algo: Make div(E)/div(B) Cleaning Optional (#2116)

* Multi-J Algo: Make div(E)/div(B) Cleaning Optional

* Remove Unnecessary Newlines

* Use IntVect Flags for F and G Index Types (#2118)

* Particle injection callback for ElectrostaticSolver (#2096)

* Particle injection callback for ElectrostaticSolver

* Move warpx_py_particleinjection to line 162, removed from OneStep_nosub

* Space above particleinjection callback

* PICMI - Embedded boundary (#2104)

* added functionality to picmi.py to allow an embedded boundary to be set from a python run script

* added test: embedded sphere run from python (the test is the same as ElectrostaticSphereEB)

* added a line to prepare_file_travis.py so that tests with USE_EB=True are not included in Python CI test list

* fixed import path in analysis.py file for new test

* added WarpX_EB flag to setup.py so that embedded boundary functionality can be included in pip installs of pywarpx

* Add particle diag to diag_store

* Make Doc Strings for the Particle Diag wrapper

* new line

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>
Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>
Co-authored-by: David Grote <grote1@llnl.gov>
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>
Co-authored-by: Luca Fedeli <luca.fedeli@cea.fr>
Co-authored-by: MaxThevenet <maxence.thevenet@desy.de>
Co-authored-by: Peter Scherpelz <31747262+peterscherpelz@users.noreply.github.com>
Co-authored-by: Edoardo Zoni <59625522+EZoni@users.noreply.github.com>
Co-authored-by: Roelof <roelof.groenewald@modernelectron.com>

EtherealMC-Bit committed:
Readded accidently removed check for self.check_timestep() for diag plotting

EtherealMC-Bit committed:
Merged memaster in

roelof-groenewald committed:
updated particle data calls for new upstream API

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -46,10 +59,35 @@ def add_field_diag(self):
+                    plotting.ArrayPlot(
+                        array=data[:, :, 0],
+                        template='rho', xaxis='z', yaxis='x', ax=ax,
+                        draw_image=True, draw_contourlines=False,
+                        cmap='viridis', plot_name="rho_" + plot_name
```
```suggestion
                        plot_name="rho_" + plot_name
```

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -46,10 +59,35 @@ def add_field_diag(self):
+                    plotting.ArrayPlot(
+                        array=data,
+                        template='phi', xaxis='z', yaxis='x', ax=ax,
+                        draw_image=True, draw_contourlines=False,
+                        cmap='viridis', plot_name="phi_" + plot_name
```
```suggestion
                        plot_name="phi_" + plot_name
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -96,8 +94,6 @@ class ArrayPlot(object):
             'legendsize': 'medium',
             'titlesize': 'large',
             'cmap': 'viridis',
```
Let's make 'viridis' the default on line 43 (I can't comment there for some reason in my current view). Then eliminate everywhere else.
```suggestion
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -111,8 +107,6 @@ class ArrayPlot(object):
             'legendsize': 'medium',
             'titlesize': 'large',
             'cmap': 'viridis',
```
```suggestion
```

peterscherpelz commented on mewarpx/mewarpx/plotting.py:
```diff
@@ -260,7 +254,8 @@ def slice_array(self):
         xaxis_idx, yaxis_idx, sliceaxis_idx, self.sliceaxis_str = (
             get_axis_idxs(self.params["xaxis"], self.params["yaxis"],
-                               self.dim))
+                            self.dim)
```
Apparently I still fail at spacing in my corrections...
```suggestion
                          self.dim)
```

peterscherpelz commented on mewarpx/mewarpx/util.py:
```diff
@@ -183,3 +184,49 @@ def J_RD(T, WF, A):
+    """Return x if x is iterable, None if x is None, [x] otherwise.
+    Useful for arguments taking either a list of single value. Strings are a
+    special case counted as 'not iterable'.
+    Arguments:
```
```suggestion

    Arguments:
```

peterscherpelz reviewed (commented):
Thanks for making lots of changes - looking through it again, a few more suggested changes + make viridis default on plotting.py:43 and in docstring at plotting.py:166 . Thanks!

EtherealMC-Bit committed:
Moved diag plotting test into /tests and fixed typo

roelof-groenewald commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -0,0 +1,93 @@
+
+            for param in self.plot_data_list:
+                if param == 'rho':
+                    data = np.array(mwxrun.get_gathered_rho_grid(include_ghosts=False)[0])
```
I think you need to do something like this here since only the root will get the data:
```suggestion
                    data = mwxrun.get_gathered_rho_grid(include_ghosts=False)
                    if mwxrun.me == 0:
                        data = np.array(data[0])
```

roelof-groenewald commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -0,0 +1,93 @@
+                        cmap='viridis', plot_name="rho_" + plot_name
+                    )
+                elif param == 'phi':
+                    data = np.array(mwxrun.get_gathered_phi_grid(include_ghosts=False)[0])
```
Same comment here as above for the `mwxrun.me` thing.

EtherealMC-Bit committed:
Made reqested changes

EtherealMC-Bit committed:
Fixed an error when running on more than one proc

EtherealMC-Bit committed:
Diag step plotting unit test

- Also includes commented out code for post processing unit test

EtherealMC-Bit committed:
Functionality and testing for plotting yt diagnostic files

roelof-groenewald commented on mewarpx/tests/test_diagnostic_plotting.py:
```diff
@@ -0,0 +1,212 @@
+
+@pytest.mark.parametrize("plot_on_diag_steps, post_processing",
+[
+    (True, False),
+    (False, True)
```
I guess with this set you could just use one and set the other one to be NOT that.

EtherealMC-Bit replied:
Using `not` on tuples or lists checks if they're empty instead of negating all the values. I could change these to lists and do:
```Python
    [True, False],
    [not e for e in [True, False]]
```
But the current way seems cleaner.

roelof-groenewald replied:
What I meant is that you can just use 1 input for the test, say `plot_on_diag_steps` and if it is false you do `post_processing`. Does that make sense?

roelof-groenewald commented on mewarpx/tests/test_diagnostic_plotting.py:
```diff
@@ -0,0 +1,212 @@
+"""
```
Two things:

1. Please name the test `test_field_diag.py` or something like that to show which file it is testing in the `mewarpx` library
2. We should make use of the diode template as much as possible, that way if an API changes in the future we won't have to manually fix several different test input files. So please do make use of the template for this test.

EtherealMC-Bit committed:
Wrapped yt loading with pytest.warns(ResourceWarning)

EtherealMC-Bit committed:
Renamed test to show what file is being tested

EtherealMC-Bit committed:
Utilized diode_setup template

EtherealMC-Bit committed:
Small cleanup

EtherealMC-Bit committed:
Fix test_custom_criteria because it causes CircleCI to fail

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -38,14 +45,18 @@ def __init__(self, diag_steps, diag_data_list, grid, name, write_dir,
         self.add_field_diag()
 
         if self.plot_on_diag_step:
             assert self.plot_data_list is not None, "plot_on_diag_step was True but plot_data_list is None!"
```
See eg https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/ for why to not use assert here
```suggestion
            if self.plot_data_list is None:
                raise ValueError(
                    "plot_on_diag_step was True but plot_data_list is None!")
```

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -58,9 +69,12 @@ def add_field_diag(self):
 
+    def add_post_processing_check_callback(self):
+        callbacks.installafterstep(self.check_for_end_of_sim)
+
```
I'm not sure we need to define separate functions in the class here, can we just install the callbacks directly from init? I don't see an advantage to the extra layer.

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -70,24 +84,65 @@ def plot_parameters(self):
+                            cmap='viridis', plot_name="phi_" + plot_name
+                        )
                 else:
                     print("Attempted to plot a parameter for which plotting has not been implemented:", param)
```
either use `warnings.warn` or raise an exception here.

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -10,9 +10,14 @@
 
+import yt
+import glob
+import pytest
```
This module shouldn't require `pytest` - that should only be wrapped in the test file itself.

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -70,24 +84,65 @@ def plot_parameters(self):
+            for datafolder in data_dirs:
+                if not "old" in datafolder:
+
+                    with pytest.warns(ResourceWarning):
```
This wrapper should be in the necessary place in the test in `tests`. pytest shouldn't touch the core code at all.

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -70,24 +84,65 @@ def plot_parameters(self):
+                        for parameter in self.diag_data_list:
+                            plot_name = parameter + "_" + datafolder.replace("diags/diags", "")
+                            data = np.mean(
+                                grid_data[parameter].to_ndarray()[:,:,0], axis=0
```
```suggestion
                                grid_data[parameter].to_ndarray()[:, :, 0], axis=0
```

peterscherpelz commented on mewarpx/mewarpx/diags_store/field_diagnostic.py:
```diff
@@ -70,24 +84,65 @@ def plot_parameters(self):
+                            plotting.ArrayPlot(
+                                    array=data,
+                                    template=template, xaxis='z', yaxis='x', ax=ax,
+                                    draw_image=True, draw_contourlines=False,
+                                    cmap='viridis', plot_name=plot_name
```
Get rid of things that are now default.
```suggestion
                                    plot_name=plot_name
```

peterscherpelz commented on mewarpx/mewarpx/diags_store/particle_diag.py:
```diff
@@ -47,7 +47,7 @@ def __init__(self, diag_steps, name=None, species=None,
 
         if self.name is None:
-            self.name= = "diag"
+            self.name= "diag"
```
```suggestion
            self.name = "diag"
```

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -130,10 +131,18 @@ class DiodeRun_V1(object):
+    # If FIELD_DIAG_PLOT_ON_DIAG_STEPS is True, what parameters to plot. Current
+    # accepted values are 'rho' and 'phi'
+    FIELD_DIAG_PLOT_DATA_LIST = None
+    # If whether or not to plot the parameters in FIELD_DIAG_DATA_LIST after
+    # the simulation is complete using yt files generated during the run during
+    # diagnostic steps.
```
```suggestion
    # Whether or not to plot the parameters in FIELD_DIAG_DATA_LIST after
    # the simulation is complete using yt files generated during the run at
    # diagnostic steps.
```

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -594,15 +603,17 @@ def init_fluxdiag(self):
-        diagnostic_intervals = f"::{self.DIAG_STEPS}"
-        self.field_diag = picmi.FieldDiagnostic(
+
+        diag = field_diagnostic.FieldDiagnostic(
```
Either save it to a member variable `self.field_diag` or don't save it at all, I think either one is OK for now, but assigning it to diag doesn't do anything.
```suggestion
        field_diagnostic.FieldDiagnostic(
```

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -11,7 +11,7 @@ def check_particle_nums():
     from pywarpx import _libwarpx
     # add 10,000 particles
-    _libwarpx.add_particles(x=np.zeros((10000)), y=np.zeros((10000)), z=np.zeros((10000)))
+    _libwarpx.add_particles(species_name="electrons", x=np.zeros((10000)), y=np.zeros((10000)), z=np.zeros((10000)))
```
`zeros` can accept an int or a tuple for a shape, but if you wanted to pass a tuple it needs to be `(10000, )`; the parentheses alone weren't doing anything.
```suggestion
    _libwarpx.add_particles(species_name="electrons", x=np.zeros(10000), y=np.zeros(10000), z=np.zeros(10000))
```

EtherealMC-Bit committed:
Merged memaster into mknr/plot_routine

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,131 @@
+    from mewarpx.setups_store import diode_setup
+    from mewarpx import testing_util
+
+    test_name = "diag_test_"
+    if plot_on_diag_steps:
+        test_name += "with_"
+    else:
+        test_name += "without_"
+    test_name += "diag_plotting_"
+    if post_processing:
+        test_name += "with_"
+    else:
+        test_name += "without_"
+    test_name += "post_processing"
```
Tons of ways to do this, maybe to make it a little more concise, do
```suggestion
    plot_diag_str = "_with_diag_plotting" if plot_on_diag_steps else ""
    post_proc_str = "_with_post_processing" if post_processing else ""
    test_name = "field_diag_test" + plot_diag_str + post_proc_str
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,131 @@
+    np.random.seed(83197410)
+
+    STEPS = 10
+    ##########################
+    # physics parameters
+    ##########################
```
Let's shorten this file
```suggestion
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,131 @@
+    FREQ = 13.56e6 # MHz
+    VOLTAGE = 450.0
+
+    ##########################
+    # numerics parameters
+    ##########################
+    # --- Grid
```
```suggestion
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,131 @@
+    number_per_cell_each_dim = [16, 32]
+    DT = 1.0 / (400 * FREQ)
+
+    # --- Number of time steps
```
```suggestion
```

peterscherpelz reviewed (changes requested):
This came together nicely! Lots of little comments but that's it.

EtherealMC-Bit committed:
Suggested changes

roelof-groenewald commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,110 @@
+"""Test the custom criteria using capacitive discharge"""
```
Note that `test_custom_criteria.py` was renamed to `test_mepicmi.py` in a different PR, so this is adding that file back to the tests. I don't think we need to do that, right?

roelof-groenewald reviewed (changes requested):
Just one comment about `test_custom_criteria.py `. Looks good otherwise.

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -545,9 +546,9 @@ def init_fluxdiag(self):
         print('### Init Diode FieldDiag ###')
 
-        diag = field_diagnostic.FieldDiagnostic(
+        self.diag = field_diagnostic.FieldDiagnostic(
```
```suggestion
        self.field_diag = field_diagnostic.FieldDiagnostic(
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,117 @@
+Monte-Carlo Collision script based on case 1 from
+Turner et al. (2013) - https://doi.org/10.1063/1.4775084
+"""
+
+from _pytest.warning_types import PytestUnraisableExceptionWarning
```
I believe this is unused.
```suggestion
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,117 @@
+import numpy as np
+
+constants = picmi.constants
+
```
```suggestion


# yt seems to be causing this error when a dataset is loaded
@pytest.mark.filterwarnings("ignore:unclosed file:ResourceWarning")
```

peterscherpelz commented on mewarpx/tests/test_field_diag.py:
```diff
@@ -0,0 +1,117 @@
+    )
+
+    while run.control.check_criteria():
+        if post_processing:
+            with pytest.warns(ResourceWarning):
+                mwxrun.simulation.step()
+        else:
+            mwxrun.simulation.step()
```
See the above `filterwarnings` suggestion.
`pytest.warns` is used to test a warning is raised when you explicitly want that warning to be raised. For us this warning is just an annoyance, so the proper behavior is to ignore it, not to check that it's raised (if yt is fixed so it doesn't happen anymore, then yay!).
```suggestion
        mwxrun.simulation.step()
```

peterscherpelz reviewed (commented):
Somehow _libwarpx did get messed up such that `get_mesh_charge_density_fp` is actually calling `libwarpx.warpx_getPhiFP`.

EtherealMC-Bit committed:
A few more suggested changes

EtherealMC-Bit committed:
Fixed issue in _libwarpx

peterscherpelz commented:
@MKieburtz Nice work! @roelof-groenewald Please make a final review and we can merge :)

roelof-groenewald reviewed (approved):
Looks good!
